tokio-util/src/io/utf8_reader.rs

```diff
@@ -0,0 +1,284 @@
+    ///
+    /// # Examples
+    /// ```
+    /// use tokio_util::io::Utf8Reader;
+    ///
+    /// # const INNER_READER: () = ();
+    /// let utf8_reader = Utf8Reader::new(INNER_READER);
+    /// ```
+    pub const fn new(inner: R) -> Self {
+        Self {
+            inner,
+            scrap_state: ScrapState::NoScrap,
+        }
+    }
+
+    /// Get the inner reader out of the `Utf8Reader`, destroying the
+    /// `Utf8Reader` in the process.
+    ///
+    /// # Examples
+    /// ```
+    /// use tokio_util::io::Utf8Reader;
+    ///
+    /// # const INNER_READER: () = ();
+    /// let utf8_reader = Utf8Reader::new(INNER_READER);
+    ///
+    /// // ... later on ...
+    /// let inner_reader = utf8_reader.into_inner();
+    /// ```
+    pub fn into_inner(self) -> R {
+        self.inner
+    }
+}
+
+impl<R> From<R> for Utf8Reader<R> {
+    fn from(inner: R) -> Self {
+        Self::new(inner)
+    }
+}
+
+impl<R: AsyncRead> AsyncRead for Utf8Reader<R> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let me = self.as_mut().project();
+        match me.scrap_state {
+            ScrapState::Scrap(scrap, len) => {
+                buf.put_slice(&scrap[..*len]);
+                *me.scrap_state = ScrapState::NoScrap;
+                self.poll_read(cx, buf)
+            }
+            ScrapState::NoScrap => {
+                ready!(me.inner.poll_read(cx, buf))?;
+
+                let filled = buf.filled();
+                let len_of_complete_utf8 = len_of_complete_or_invalid_utf8_bytes(filled);
+                if len_of_complete_utf8 != filled.len() {
+                    // should not be greater than 3, as
+                    // `len_of_complete_or_invalid_utf8_bytes` should only ever
+                    // return a length that's at most 3 less than the length of
+                    // the byte-string
+                    let scrap_len = filled.len() - len_of_complete_utf8;
+                    let mut scrap = [0; 3];
+                    scrap[..scrap_len].copy_from_slice(&filled[len_of_complete_utf8..]);
+                    *me.scrap_state = ScrapState::Scrap(scrap, scrap_len);
+                    // shouldn't panic, because `len_of_complete_utf8` will be less than or
+                    // equal to `filled.len()`, which is guarenteed to be less than the
+                    // initialized portion of `buf`
+                    buf.set_filled(len_of_complete_utf8);
+                }
+
+                Poll::Ready(Ok(()))
+            }
+        }
+    }
+}
+
+impl<R: AsyncBufRead> AsyncBufRead for Utf8Reader<R> {
+    fn poll_fill_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<&[u8]>> {
+        let slice = ready!(self.project().inner.poll_fill_buf(cx))?;
+        Poll::Ready(Ok(&slice[..len_of_complete_or_invalid_utf8_bytes(slice)]))
+    }
```

Comment:
If `len_of_complete_or_invalid_utf8_bytes` returns zero here for a slice that is non-empty, then that is incorrect. Returning an empty slice from `poll_fill_buf` indicates that the stream has reached its end and that no bytes will ever be yielded again.

Reply:
In that case, we return `Poll::Pending`, right?

Reply:
No, you would never be woken up again as the `poll_fill_buf` call on the inner IO resource didn't return `Poll::Pending`. You have to somehow store the partial character, then read more to get a full character.

Reply:
It shouldn't be necessary to store it, as the underlying `AsyncBufRead` will keep those bytes in its buffer until `consume` is called, right? In which case, we can just `return self.poll_fill_buf(cx)` to try again, right?

Reply:
Yes, but the only way to get it to read more data is to consume the ones that are currently available.

Reply:
Having thought about it, I don't know how that would work. If I store bytes somewhere, how do I put them in front of new bytes obtained from `poll_fill_buf`? I can do that with `poll_read`, but since `poll_fill_buf` returns an immutable slice, unless I maintain my own buffer and copy everything from the inner `AsyncBufRead`, I can't do something like that. Should `Utf8Reader` just not implement `AsyncBufRead`?

Reply:
To be clear, `Utf8Reader` *could* maintain its own buffer, but that would obviously be expensive.

Reply:
You don't need to have a large buffer. You can just fill the buffer up until it contains an entire character, then provide only that even if the underlying IO resource has more available. The user would need to consume the character to get to the rest of the data.